packages/ai-jsx/src/batteries/constrained-output.tsx

```diff
@@ -291,23 +293,15 @@ export async function JsonChatCompletionFunctionCall(
     >
       {children}
       <SystemMessage>
-        Your response must use the `print` function that is provided. No other explanation needed.
+        Your response must use the `print` function that is provided. No other explanation needed. do not respond with
+        an assistant message. Just call the function.
       </SystemMessage>
     </ChatCompletion>
   );
 
-  const output = await render(childrenWithCompletion, { stop: (e) => e.tag == FunctionCall });
-
-  // TODO: can we make this work with progressive loading? This will require changes to core/completion.
-
-  const lastYield = output[output.length - 1];
-  if (AI.isElement(lastYield) && lastYield.tag == FunctionCall) {
-    // TODO: is validation necessary? If yes, how to handle errors?
-    schema.parse(lastYield.props.args);
-    return JSON.stringify(lastYield.props.args);
+  const frames = render(childrenWithCompletion);
+  for await (const frame of frames) {
+    yield JSON.stringify(JSON.parse(frame).arguments);
   }
-  // TODO: retry maybe?
-  throw new AIJSXError('Expected a function call to `print`.', ErrorCode.ModelOutputDidNotMatchConstraint, 'runtime', {
-    output: JSON.stringify(output),
-  });
+  return JSON.stringify(JSON.parse(await frames).arguments);
```

Comment:
I'm not a fan of how we have so many `JSON.parse` and `JSON.stringify`'s, but this is okay for now.

Reply:
Yeah it's not ideal – feel free to change it if you have a better suggestion. :smile: